### PR TITLE
changes due to controller-runtime v0.11.0 upgrade

### PR DESCRIPTION
**Pull request by vijtrip2**
Issue #, if available: https://github.com/aws-controllers-k8s/community/issues/1112

Description of changes:
* Upgrade of controller-runtime to v0.11.0
* Upgrade of go.mod to 1.17 because that is minimal go version for controller-runtime library
*  Keeping the indirect import in go.mod file because removing them causes the error "go.mod needs changes. Run go mod tidy" during local development
> At go 1.17 and above, the go command adds an indirect requirement for each module that provides any package imported (even indirectly) by a package or test in the main module or passed as an argument to go get. These more comprehensive requirements enable module graph pruning and lazy module loading.
* `Reconcile` method in reconcilers now accepts a `context.Context` parameter
* Replace `ACK RuntimeMetaObject` with `controller-runtime Object`
* Replace `desired.RuntimeObject().DeepCopyObject()` with `desired.DeepCopy().RuntimeObject()` because the former does not return implementation of `controller-runtime Object`
* Remove `RuntimeMetaObject()` method from AWSResource interface
* Use `controller-runtime client Object` instead of `apimachinery runtime Object`

------

* Tested locally by running e2e tests for ecr-controller
* Validated that newly generated image has no golang security vulnerabilities.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by vijtrip2** [comment]
note: unit test is currently failing because prow base image for testing the controller is not using `go1.17`. working on fixing it

**Comment by jaypipes** [comment]
> * Replace `ACK RuntimeMetaObject` with `controller-runtime Object`

:+1: :tada:

File: go.mod [hunk 1]
```diff
@@ -1,19 +1,67 @@
...
-	k8s.io/api v0.18.2
-	k8s.io/apimachinery v0.18.6
-	k8s.io/client-go v0.18.2
-	sigs.k8s.io/controller-runtime v0.6.0
+	github.com/stretchr/testify v1.7.0
+	go.uber.org/zap v1.19.1
+	k8s.io/api v0.23.0
+	k8s.io/apimachinery v0.23.0
+	k8s.io/client-go v0.23.0
+	sigs.k8s.io/controller-runtime v0.11.0
```
Comment (RedbackThomson) [inline go.mod:18]: I'd love to see what else we unlocked in each of these version updates. Especilaly `logr` going from `0.1.0` to `1.2.0` - I imagine there are some interesting features we might use.

File: pkg/runtime/tags.go [hunk 2]
```diff
@@ -18,15 +18,15 @@ import (
 	"time"
 
 	ackconfig "github.com/aws-controllers-k8s/runtime/pkg/config"
-	acktypes "github.com/aws-controllers-k8s/runtime/pkg/types"
+	k8sctrlrtclient "sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (RedbackThomson) [inline pkg/runtime/tags.go:21]: This is quite a mouthful haha. Perhaps just `rtclient` instead?

**Review by RedbackThomson — APPROVED** [review]
So excited for these changes. Only took 2 years for this to be merged, but I'm glad we got there.

**Review by jaypipes — APPROVED** [review]
I like @RedbackThomson's naming suggestion too

File: pkg/runtime/reconciler_test.go [hunk 3]
```diff
@@ -197,7 +195,10 @@ func TestReconcilerUpdate(t *testing.T) {
 
 	r, kc := reconcilerMocks(rmf)
 
-	kc.On("Patch", ctx, latestRTObj, client.MergeFrom(desiredRTObj)).Return(nil)
+	// pointers returned from "client.MergeFrom" fails the equality check during
+	// assertion even when parameters inside two objects are same.
+	// hence we use mock.AnythingOfType parameter to assert patch call
+	kc.On("Patch", ctx, latestRTObj, mock.AnythingOfType("*client.mergeFromPatch")).Return(nil)
```
[the remark this answers is not in the document]
Comment (vijtrip2) [reply pkg/runtime/reconciler_test.go:201]: AFAIKT, It just performs a string match

**Comment by RedbackThomson** [comment]
Let's do it!
/lgtm

**Pull request merged**